enhancements/ingress/set-delete-http-headers.md

```diff
@@ -0,0 +1,1120 @@
+```
+G) Dynamic values are allowed. (more details can be found in https://cbonte.github.io/haproxy-dconv/2.4/configuration.html#7.3) 
+Example dynamic values for request:
+```conf
+%[req.hdr(host),lower]
+%[req.hdr(host),base64]
+%[req.hdr(Host)]
+```
+Example dynamic values for response:
+```conf
+%[res.hdr(X-Acl-Key),lower]
+%[res.hdr(X-Acl-Key),base64]
+%[req.hdr(Host)]
+```
+
+Example dynamic values that can be used by both request and response:
+```conf
+%{+Q}[ssl_c_der,base64]
+```
+
+H) Static header values
+For regular header which are not dynamic one can pass values as mentioned in unit test `TestHeaderValues` with `validInput:true`
+present in `pkg/route/validation/validation_test.go`
+
+Please note that the router quotes all values before interpolating them into `haproxy.conf`.
+
+The user must not add quotes in the API `value` field unless the user wants literal quotes in the HTTP header value.
+Specifically, any single-quote present in the value is escaped, and then the whole value
+is surrounded in single-quotes.  This prevents HAProxy config parsing errors in the event of invalid input. 
+
+This quoting logic will be present in the router.
+
+To quote using single-quotes, we will do the following:
+1. Replace `'` with `'\''`.
+2. Wrap the entire string in single quotes.
+   For example, `"abc"\\ 'def'`  will be converted to `'"abc"\\ '\''def'\'''`.
+
+
+### Risks and Mitigations
+
+We identified the following headers that require special explanations:
+
+#### Proxy header
+The `proxy` HTTP request header can be used to exploit vulnerable CGI applications using the [_httpoxy_](https://httpoxy.org/) vulnerability.
+In this vulnerability, if the client sets the HTTP `proxy` request header and the proxy passes it through to the application, then the CGI application takes the header value and injects it in the `HTTP_PROXY` environment variable.
+
+In this manner, an attacker could direct a vulnerable application to use a proxy of the attacker's choice.
+Note that OpenShift router already mitigates this vulnerability by deleting any `proxy` header that the client specifies.
+
+Allowing the cluster-admin or route owner to override the `proxy` HTTP header isn't necessarily a security risk; it would only allow admins or route owners to inject configuration into their own applications, which they already control.
+However, the `proxy` header is non-standard, and allowing it to be configured is error-prone.
+
+For this reason, we have not and still do not allow setting the `proxy` header
+
+#### Strict-Transport-Security
+Strict-transport-security is already [handled separately](https://docs.openshift.com/container-platform/4.13/networking/routes/route-configuration.html#nw-enabling-hsts_route-configuration),
+so we don't allow overrides of that configuration.
+
+We don't allow setting the `strict-transport-security` headers on either ingress or route.
```

Comment:
You mean IngressController, not Ingress, right?  
```suggestion
We don't allow setting the `strict-transport-security` header on either an IngressController or a Route.
```

Reply:
yes right